<a href="https://colab.research.google.com/github/DunieCarrasco/deep-learning-challenge/blob/main/Starter_Code.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Preprocessing

In [2]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(["EIN", "NAME"], axis=1)

In [4]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [5]:
# Look at APPLICATION_TYPE value counts for binning
application_df['APPLICATION_TYPE'].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
cutoff = 500
app_counts = application_df['APPLICATION_TYPE'].value_counts()
application_types_to_replace = list(app_counts[app_counts <= cutoff].index)

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [7]:
# Look at CLASSIFICATION value counts for binning
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [8]:
# You may find it helpful to look at CLASSIFICATION value counts >1
class_counts = application_df['CLASSIFICATION'].value_counts()
class_counts_gt1 = class_counts[class_counts > 1]
print(class_counts_gt1)

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64


In [9]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
cutoff = 1000
app_counts = application_df['CLASSIFICATION'].value_counts()
classifications_to_replace = list(app_counts[app_counts <= cutoff].index)

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [10]:
application_df.columns

Index(['APPLICATION_TYPE', 'AFFILIATION', 'CLASSIFICATION', 'USE_CASE',
       'ORGANIZATION', 'STATUS', 'INCOME_AMT', 'SPECIAL_CONSIDERATIONS',
       'ASK_AMT', 'IS_SUCCESSFUL'],
      dtype='object')

In [11]:
application_df.dtypes

APPLICATION_TYPE          object
AFFILIATION               object
CLASSIFICATION            object
USE_CASE                  object
ORGANIZATION              object
STATUS                     int64
INCOME_AMT                object
SPECIAL_CONSIDERATIONS    object
ASK_AMT                    int64
IS_SUCCESSFUL              int64
dtype: object

In [12]:
# Convert categorical data to numeric with `pd.get_dummies`
get_dummies_application_df = pd.get_dummies(application_df, columns=['APPLICATION_TYPE', 'AFFILIATION', 'CLASSIFICATION', 'USE_CASE',
       'ORGANIZATION', 'INCOME_AMT', 'SPECIAL_CONSIDERATIONS'])

In [13]:
# Split our preprocessed data into our features and target arrays
y = get_dummies_application_df["IS_SUCCESSFUL"].values
X = get_dummies_application_df.drop(["IS_SUCCESSFUL"],1).values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

<ipython-input-13-a509f3408616>:3: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X = get_dummies_application_df.drop(["IS_SUCCESSFUL"],1).values


In [14]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [29]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_dim=X.shape[1]

nn_model = tf.keras.models.Sequential()
dense = tf.keras.layers.Dense(input_dim=input_dim, units=80, activation="relu")
nn_model.add(dense)

# First hidden layer
dense_1 = tf.keras.layers.Dense(units=30, activation="relu")
nn_model.add(dense_1)

# Second hidden layer
dense_2 = tf.keras.layers.Dense(units=30, activation="tanh")
nn_model.add(dense_2)

# Output layer
output_layer = tf.keras.layers.Dense(units=1, activation="sigmoid")
nn_model.add(output_layer)

# Check the structure of the model
nn_model.summary()

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_14 (Dense)            (None, 80)                3520      
                                                                 
 dense_15 (Dense)            (None, 30)                2430      
                                                                 
 dense_16 (Dense)            (None, 1)                 31        
                                                                 
Total params: 5,981
Trainable params: 5,981
Non-trainable params: 0
_________________________________________________________________


In [26]:
# Compile the model
nn_model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [27]:
# Train the model
fit_model = nn_model.fit(X_train_scaled, y_train, epochs=50)


Epoch 1/50
804/804 [==============================] - 47s 3ms/step - loss: 0.5690 - accuracy: 0.7233
Epoch 2/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5550 - accuracy: 0.7307
Epoch 3/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5521 - accuracy: 0.7324
Epoch 4/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5507 - accuracy: 0.7339
Epoch 5/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5496 - accuracy: 0.7331
Epoch 6/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5491 - accuracy: 0.7329
Epoch 7/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5477 - accuracy: 0.7338
Epoch 8/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5476 - accuracy: 0.7356
Epoch 9/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5474 - accuracy: 0.7341
Epoch 10/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5469 - accuracy: 0.733

In [28]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5543 - accuracy: 0.7262 - 477ms/epoch - 2ms/step
Loss: 0.5542869567871094, Accuracy: 0.7261807322502136


In [24]:
# Export our model to HDF5 file
nn_model.save("AlphabetSoupCharity.h5")

**Overview**

The purpose of this analysis is to develop a machine learning model to select the a fictitious charitable organization for funding with the best chance of success in their ventures. The dataset from the nonprofit foundation Alphabet Soup containing more than 34,000 organizations that have received funding from Alphabet Soup over the years. That data contains information about application type, affiliation, classification, use case for funding, organization type, active status, income, special considerations, funding amount requested and if it was considered successful for each applicant.

The goal is to use this data to develop a model that can accurately predict which charitable organization will be successful, in order to help Alphabet Soup make more informed decisions about which charitable organization applications to approve. Specifically, we will be using a binary classification model to predict whether the funding will be approved or not.

The analysis involves several steps, including data cleaning and preprocessing, feature engineering, model selection and training, and evaluation of model performance. The end goal is to develop a model that can accurately predict funding success based on the available data, and to use this model to inform future funding approval decisions.

**Results**
* Data Preprocessing
 * What variable(s) are the target(s) for your model?
   - My target in my model is IS_SUCCESSFUL data 
 * What variable(s) are the features for your model?
   - My variables in my model is 'APPLICATION_TYPE', 'AFFILIATION', 'CLASSIFICATION', 'USE_CASE',
       'ORGANIZATION', 'STATUS', 'INCOME_AMT', 'SPECIAL_CONSIDERATIONS',
       'ASK_AMT' data.
 * What variable(s) should be removed from the input data because they are neither targets nor features?
   - Variable that I removed from the data where "EIN", "NAME" as they were neither targets nor features.

* Compiling, Training, and Evaluating the Model
 * How many neurons, layers, and activation functions did you select for your neural network model, and why?
   - For the neural network model developed for this classification problem, I selected a total of four layers, with 80 neurons in the first layer, 30 neurons in the second layer, and 10 neurons in both the third and output layers.
 * Were you able to achieve the target model performance?
   - I was not able to achieve the target goal of above 75% accuracy.

**Summary**

The deep learning model developed for this classification problem achieved an accuracy score of approximately 72%. While this is better than random guessing, it is not a particularly high accuracy score and there is likely room for improvement.

One potential approach for improving the model's performance is to try a different algorithm, such as a Random Forest model. These algorithms have been shown to perform well in a wide range of classification problems and may be better suited to the particular features of this dataset.